midi/drivers/coremidi.c

```diff
@@ -0,0 +1,415 @@
+    }
+}
+
+/// Initialize the CoreMIDI client and ports
+static void *coremidi_init(const char *input, const char *output) {
+    coremidi_t *d = (coremidi_t *)calloc(1, sizeof(coremidi_t));
+    if (!d) {
+        RARCH_ERR("[MIDI]: Out of memory.\n");
+        return NULL;
+    }
+
+    OSStatus err = MIDIClientCreate(CFSTR("RetroArch MIDI Client"), NULL, NULL, &d->client);
+    if (err != noErr) {
+        RARCH_ERR("[MIDI]: MIDIClientCreate failed: %d\n", err);
+        free(d);
+        return NULL;
+    } else {
+        RARCH_LOG("[MIDI]: CoreMIDI client created successfully\n");
+    }
+
+    // Create input port if specified
+    if (input) {
+        err = MIDIInputPortCreate(d->client, CFSTR("Input Port"), midi_read_callback, d, &d->input_port);
+        if (err != noErr) {
+            RARCH_ERR("[MIDI]: MIDIInputPortCreate failed: %d\n", err);
+            MIDIClientDispose(d->client);
+            free(d);
+            return NULL;
+        }
+    } else {
+        RARCH_LOG("[MIDI]: CoreMIDI input port created successfully\n");
+    }
+
+    // Create output port if specified
+    if (output) {
+        err = MIDIOutputPortCreate(d->client, CFSTR("Output Port"), &d->output_port);
+        if (err != noErr) {
+            RARCH_ERR("[MIDI]: MIDIOutputPortCreate failed: %d\n", err);
+            if (d->input_port)
+                MIDIPortDispose(d->input_port);
+            MIDIClientDispose(d->client);
+            free(d);
+            return NULL;
+        }
+    } else {
+        RARCH_LOG("[MIDI]: CoreMIDI output port created successfully\n");
+    }
+
+    return d;
+}
+
+/// Get available input devices
+static bool coremidi_get_avail_inputs(struct string_list *inputs) {
+    ItemCount count = MIDIGetNumberOfSources();
+    union string_list_elem_attr attr = {0};
+
+    for (ItemCount i = 0; i < count; i++) {
+        MIDIEndpointRef endpoint = MIDIGetSource(i);
+        CFStringRef name;
+        char buf[256];
+
+        OSStatus err = MIDIObjectGetStringProperty(endpoint, kMIDIPropertyDisplayName, &name);
+        if (err == noErr) {
+            if (CFStringGetCString(name, buf, sizeof(buf), kCFStringEncodingUTF8)) {
+                if (!string_list_append(inputs, buf, attr)) {
+                    RARCH_ERR("[MIDI]: Failed to append input device to list: %s\n", buf);
+                    CFRelease(name);
+                    return false;
+                } else {
+                    RARCH_LOG("[MIDI]: Input device added to list: %s\n", buf);
+                }
+            }
+            CFRelease(name);
+        } else {
+            RARCH_WARN("[MIDI]: Failed to get display name for input device %d: %d\n", i, err);
+        }
+    }
+
+    return true;
+}
+
+/// Get available output devices
+static bool coremidi_get_avail_outputs(struct string_list *outputs) {
+    ItemCount count = MIDIGetNumberOfDestinations();
+    union string_list_elem_attr attr = {0};
+
+    for (ItemCount i = 0; i < count; i++) {
```

Comment:
Incompatible with c89

Reply:
I'll fix it up later, we can still merge it as is.